File: Prove/Python/01 - EsercitazioneOOP-Concorrenza/prod_cons_CV/package_coda/codaWrapperLock.py
from .coda import coda
import threading


class codaWrapperLock(coda):
    
    def __init__(self,size):
        super().__init__(size)
        self.lock = threading.RLock()
        self.MSG_DISP = threading.Condition(lock=self.lock)
        self.SPZ_DISP = threading.Condition(lock=self.lock)

    def inserisci(self, elem):

        with self.lock: #lock().acquire()
        
            while(super().full()):
                self.SPZ_DISP.wait()
            
            super().inserisci(elem=elem)

            self.MSG_DISP.notify()

        #lock().release
        
       
    

    def preleva(self):

        with self.lock:

            while(super().empty()):
                self.MSG_DISP.wait()
            
            elem = super().preleva()
            self.SPZ_DISP.notify()
            return elem
        
        
    

File: Java/04 - Proxy Skeleton/Esercitazione Magazzino2/src/coda/CodaWrapperLock.java
package coda;

import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.ReentrantLock;

public class CodaWrapperLock extends CodaWrapper {
    
    private ReentrantLock lock=null;
    private Condition SPZ_DISP=null;
    private Condition MSG_DISP=null;


    public CodaWrapperLock(ICoda c){
        super(c);
        lock = new ReentrantLock();
        SPZ_DISP = lock.newCondition();
        MSG_DISP = lock.newCondition();
    }

    public void inserisci(int id){

        try{
            lock.lock();

            while(coda.full()){
                
                try{
                    SPZ_DISP.await();
                    
                }catch(InterruptedException ie){
                    System.err.println("Errore nella wait su SPZ_DISP");
                }
            }
            
            coda.inserisci(id);
            MSG_DISP.signal();

        }finally{
            lock.unlock();
        }

    }

    public int preleva(){
        int val=-1;
        try{
            lock.lock();
      
            while (coda.empty()) {
                try{
                    MSG_DISP.await();
                }catch(InterruptedException ie){
                    System.err.println("Errore nella wait su SPZ_DISP");
                }
            }

            val = coda.preleva();

        }finally{
            lock.unlock();
        }
        
        return val;

    }



}
